lighthouse-core/gather/gatherers/dom-timeline.js

```diff
@@ -0,0 +1,139 @@
+      for (const node of addedNodes) {
+        if (node.nodeName !== 'IFRAME') continue;
+        // TODO: verify that Iframe src is of an ad network / ignore non-ad iframes
+        // @ts-expect-error ___observedIframes does not exist on window by default
+        window.___observedIframes.push({
+          time: currTime,
+          // @ts-expect-error - put into scope via stringification
+          devtoolsNodePath: getNodePath(node),
+          // @ts-expect-error - put into scope via stringification
+          snippet: getOuterHTMLSnippet(node),
+          // @ts-expect-error - put into scope via stringification
+          selector: getNodeSelector(node),
+          // @ts-expect-error - put into scope via stringification
+          nodeLabel: getNodeLabel(node),
+        });
+      }
+    }
+  });
+  // @ts-expect-error ___observer does not exist on window by default
+  window.___observer.observe(document, {childList: true, subtree: true});
+}
+
+/**
+ * //@return {Array<LH.Artifacts.DOMTimestamp>}
+ */
+function getDOMTimestamps() {
+  // @ts-expect-error ___observer does not exist on window by default
+  window.___observer.disconnect();
+  // @ts-expect-error ___observedIframes does not exist on window by default
+  return window.___observedIframes;
+}
+
+/**
+ * @param {LH.Artifacts.TraceOfTab['layoutShiftTimelineEvents']} layoutEvents
+ * @return {Array<LH.Artifacts.DOMWindow>}
+ */
+function getLayoutShiftWindows(layoutEvents) {
+  const windows = [];
+  let end = undefined;
+  let start = undefined;
+  for (let i = layoutEvents.length - 1; i >= 0; i--) {
+    if (!end && layoutEvents[i].event.name === 'LayoutShift') {
+      if (i - 1 >= 0 && layoutEvents[i - 1].event.name === 'UpdateLayerTree') {
+        // TODO: have to confirm that the layoutshift happens within ULT
+        end = layoutEvents[i - 1].timing;
+        i -= 1;
+        continue;
+      }
+    }
+    if (end && layoutEvents[i].event.name === 'UpdateLayerTree') {
+      start = layoutEvents[i].timing;
+      windows.push({start, end});
+      end = undefined;
+      start = undefined;
+    } else if (end && layoutEvents[i].event.name === 'LayoutShift') {
+      if (i - 1 >= 0 && layoutEvents[i - 1].event.name === 'UpdateLayerTree') {
+        start = layoutEvents[i - 1].timing;
+        windows.push({start, end});
+        end = layoutEvents[i - 1].timing;
+        start = undefined;
+        i -= 1;
+      }
+    }
+  }
+  return windows;
+}
+
+class DOMTimeline extends Gatherer {
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   */
+  // @ts-expect-error we will not return a phase result
+  async beforePass(passContext) {
+    const script = `(() => {
+      ${pageFunctions.getNodePathString}
+      ${pageFunctions.getNodeLabelString}
+      ${pageFunctions.getNodeSelectorString}
+      ${pageFunctions.getOuterHTMLSnippetString}
+      ${setupObserver.toString()
+        .replace('function setupObserver() {', '')
+        .replace('e});\n}', 'e});')}
```

Comment:
we would typically just do
```suggestion
      return (${setupObserver.toString()})();
```

:)